{% load static %}  <!-- Add this line to load static template tag -->
{% if output.flag == 'warning' %}
  <div class="row">
    <div class="col-md-12 mb-5">
    {% if "Protein level variant descriptions are not fully supported due to redundancy in the genetic code" in output.warnings %}
      <h1>Protein variant warnings</h1>
      <ul>
    {% endif %}
    {% if "Protein level variant descriptions are not fully supported due to redundancy in the genetic code" not in output.warnings %}
      <h1>Error</h1>
      <p>Unable to validate the submitted variant <code>{{ output.inputted }}</code> against the {{ output.genome }} assembly. The following warnings were returned:</p>
      <ul>
     {% endif %}
      {% for warning in output.warnings %}
        <li class="text-danger">{{ warning }}</li>
      {% endfor %}
      </ul>
      <p>
        {% if 'does not agree with reference sequence' in output.warnings.0 %}
          Based on the error above you likely selected the wrong assembly.
        {% endif %}
      </p>
    </div>
  </div>
{% else %}
  {% if output.results|length > 1 %}
      <div class="row">
          <div class="col-md-12 mb-5">
              <p>VariantValidator returned multiple results, select those you are interested in from the table below</p>
              <table class="table table-bordered text-center">
                  <tr>
                      <th>Select</th>
                      <th>Transcript Accession</th>
                      <th>Gene Accession</th>
                      <th>Latest Version</th>
                      <th>MANE Select</th>
                      <th>MANE Plus Clinical</th>
                      <th>RefSeq Select</th>
                  </tr>
                  {% for each in output.results %}
                      <tr>
                          <td>
                              <div class="custom-control custom-checkbox">
                                  <input type="checkbox" class="custom-control-input res-checkbox" data-tabid="{{ each.id }}" id="{{ each.id }}-checkbox">
                                  <label class="custom-control-label" for="{{ each.id }}-checkbox"></label>
                              </div>
                          </td>
                          <td>{{ each.tx_ac }}</td>
                          <td>{{ each.gene_ac }}</td>
                          <td class="text-center">
                              {% if each.latest %}
                                  <img src="{% static 'img/tick-green-20x15.png' %}" alt="Tick" width="20" height="15">
                              {% else %}
                                  <img src="{% static 'img/cross-red-15x15.png' %}" alt="Cross" width="15" height="15">
                              {% endif %}
                          </td>
                          <td class="text-center">
                              {% if each.annotations.mane_select %}
                                  <img src="{% static 'img/tick-green-20x15.png' %}" alt="Tick" width="20" height="15">
                              {% else %}
                                  <img src="{% static 'img/cross-red-15x15.png' %}" alt="Cross" width="15" height="15">
                              {% endif %}
                          </td>
                          <td class="text-center">
                              {% if each.annotations.mane_plus_clinical %}
                                  <img src="{% static 'img/tick-green-20x15.png' %}" alt="Tick" width="20" height="15">
                              {% else %}
                                  <img src="{% static 'img/cross-red-15x15.png' %}" alt="Cross" width="15" height="15">
                              {% endif %}
                          </td>
                          <td class="text-center">
                              {% if each.annotations.refseq_select %}
                                  <span><i class="fas fa-check text-success"></i></span>
                              {% else %}
                                  <span><i class="fas fa-times text-danger"></i></span>
                              {% endif %}
                          </td>
                      </tr>
                  {% endfor %}
              </table>
          </div>
      </div>
  {% endif %}
  <div id="results">
      <div class="row">
          <div class="col-md-12 mb-2">
              <ul class="nav nav-tabs" id="myTab" role="tablist">
                  {% for each in output.results %}
                      <li class="nav-item">
                          <a class="nav-link" id="{{ each.id }}-tab" data-toggle="tab" href="#{{ each.id }}" role="tab" aria-controls="{{ each.id }}" aria-selected="false">{{ each.safe_hgvs_trans }}</a>
                      </li>
                  {% endfor %}
              </ul>
              <div class="tab-content" id="myTabContent">
                  {% for each in output.results %}
                      <div class="tab-pane" id="{{ each.id }}" role="tabpanel" aria-labelledby="{{ each.id }}-tab">
                          <h3 class="mt-3">{{ each.safe_hgvs_trans }}</h3>
                          <h4>{{ each.transcript_description }}</h4>
                          <ul>
                          {% if each.annotations.map %}
                              <li><h5>Selected genome build: {{ output.genome }}</h5></li>
                              <li><h5>Map location: {{ each.annotations.map }}</h5></li>
                          {% endif %}
                          {% if not each.annotations.map %}
                              <li><h5>Selected genome build: {{ output.genome }}</h5></li>
                              <li><h5>Map location: {{ each.chr_loc }}</h5></li>
                          {% endif %}
                          {% if each.annotations.db_xref.select %}
                              {% if "Plus" in each.annotations.db_xref.select %}
                                  <li><h5>Transcript Flag: {{ each.annotations.db_xref.select}}</h5></li>
                              {% endif %}
                              {% if not "Plus" in each.annotations.db_xref.select %}
                                  <li><h5>Transcript Flag: {{ each.annotations.db_xref.select}} Select</h5></li>
                              {% endif %}
                          {% endif %}
                          {% if each.annotations.db_xref.CCDS%}
                              <li><h5>CCDS ID: <a href="https://www.ncbi.nlm.nih.gov/CCDS/CcdsBrowse.cgi?REQUEST=CCDS&DATA={{ each.annotations.db_xref.CCDS }}" target="_blank">{{ each.annotations.db_xref.CCDS }}</a></h5></li>
                          {% endif %}
                          </ul>
                          <br>
                          {% if each.validation_warnings and not each.rna_variant_descriptions.usage_warnings %}
                          <h5>Warnings output during validation</h5>
                            <ul>
                            {% for warning in each.validation_warnings %}
                              <li class="text-danger">{{ warning }}</li>
                            {% endfor %}
                            </ul>
                          {% endif %}
                          {% if each.rna_variant_descriptions.usage_warnings%}
                          <h5>Warnings output during validation</h5>
                            <ul>
                            {% for warning in each.rna_variant_descriptions.usage_warnings %}
                              <li class="text-danger">{{ warning }}</li>
                            {% endfor %}
                            </ul>
                          {% endif %}
                          <h4>HGVS-compliant variant descriptions</h4>
                          <table class="table table-bordered">
                              <tr>
                                  <th>Type</th>
                                  <th>Variant Description</th>
                                  <th>Link to Reference sequence Record</th>
                              </tr>
                              {% if each.hgvs_transcript_variant %}
                              <tr>
                                  <td>Transcript (:c.)</td>
                                  <td><a class="variant-example" data-genome="{{ output.genome }}" href="#">{{ each.hgvs_transcript_variant }}</a></td>
                                  <td><a href="https://www.ncbi.nlm.nih.gov/nuccore/{{ each.tx_ac }}" target="_blank">{{ each.tx_ac }}</a></td>
                              </tr>
                              {%  endif %}

                              {% if each.rna_variant_descriptions.rna_variant%}
                              <tr>
                                  <td>Transcript (:r.)</td>
                                  <td><a class="variant-example" data-genome="{{ output.genome }}" href="#">{{ each.rna_variant_descriptions.rna_variant }}</a></td>
                                  <td><a href="https://www.ncbi.nlm.nih.gov/nuccore/{{ each.tx_ac }}" target="_blank">{{ each.tx_ac }}</a></td>
                              </tr>
                              {%  endif %}

                              {% if each.hgvs_lrg_transcript_variant %}
                              <tr>
                                  <td>Transcript (:c.)</td>
                                  <td><a class="variant-example" data-genome="{{ output.genome }}" href="#">{{ each.hgvs_lrg_transcript_variant}}</a></td>
                                  <td><a href="{{ each.reference_sequence_records.lrg }}" target="_blank">{{ each.lrg_ac }}</a></td>
                              </tr>
                              {% endif %}
                              {% if each.hgvs_refseqgene_variant %}
                              <tr>
                                  <td>RefSeq Gene (:g.)</td>
                                  <td><a class="variant-example" data-genome="{{ output.genome }}" href="#">{{ each.hgvs_refseqgene_variant }}</a></td>
                                  <td><a href="https://www.ncbi.nlm.nih.gov/nuccore/{{ each.gene_ac }}" target="_blank">{{ each.gene_ac }}</a></td>
                              </tr>
                              {% endif %}
                              {% if each.hgvs_lrg_variant %}
                              <tr>
                                  <td>LRG (:g.)</td>
                                  <td><a class="variant-example" data-genome="{{ output.genome }}" href="#">{{ each.hgvs_lrg_variant }}</a></td>
                                  <td><a href="{{ each.reference_sequence_records.lrg }}" target="_blank">{{ each.lrg_ac }}</a></td>
                              </tr>
                              {% endif %}
                              {% if each.hgvs_predicted_protein_consequence.tlr %}
                              <tr>
                                  <td>Protein (:p.)</td>
                                  <td><a class="variant-example" data-genome="{{ output.genome }}" href="#">{{ each.hgvs_predicted_protein_consequence.tlr }}</a></td>
                                  <td><a href="https://www.ncbi.nlm.nih.gov/nuccore/{{ each.prot_ac }}" target="_blank">{{ each.prot_ac }}</a></td>
                              </tr>
                             {% endif %}
                              {% if each.hgvs_predicted_protein_consequence.slr %}
                              <tr>
                                  <td>Protein (:p.)</td>
                                  <td><a class="variant-example" data-genome="{{ output.genome }}" href="#">{{ each.hgvs_predicted_protein_consequence.slr }}</a></td>
                                  <td><a href="https://www.ncbi.nlm.nih.gov/nuccore/{{ each.prot_ac }}" target="_blank">{{ each.prot_ac }}</a></td>
                              </tr>
                             {% endif %}

                              {% if each.rna_variant_descriptions.translation  %}
                              <tr>
                                  <td>Protein (:p.) translated from r.</td>
                                  <td><a class="variant-example" data-genome="{{ output.genome }}" href="#">{{ each.rna_variant_descriptions.translation }}</a></td>
                                  <td><a href="https://www.ncbi.nlm.nih.gov/nuccore/{{ each.prot_ac }}" target="_blank">{{ each.prot_ac }}</a></td>
                              </tr>
                             {% endif %}
                              {% if each.rna_variant_descriptions.translation_slr %}
                              <tr>
                                  <td>Protein (:p.) translated from r.</td>
                                  <td><a class="variant-example" data-genome="{{ output.genome }}" href="#">{{ each.rna_variant_descriptions.translation_slr }}</a></td>
                                  <td><a href="https://www.ncbi.nlm.nih.gov/nuccore/{{ each.prot_ac }}" target="_blank">{{ each.prot_ac }}</a></td>
                              </tr>
                             {% endif %}

                              {% if each.hgvs_predicted_protein_consequence.lrg_tlr  %}
                              <tr>
                                  <td>LRG Protein (:p.)</td>
                                  <td><a class="variant-example" data-genome="{{ output.genome }}" href="#">{{ each.hgvs_predicted_protein_consequence.lrg_tlr }}</a></td>
                                  <td><a href="{{ each.reference_sequence_records.lrg }}" target="_blank">{{ each.lrg_ac }}</a></td>
                              </tr>
                             {% endif %}
                              {% if each.hgvs_predicted_protein_consequence.lrg_slr  %}
                              <tr>
                                  <td>LRG Protein (:p.)</td>
                                  <td><a class="variant-example" data-genome="{{ output.genome }}" href="#">{{ each.hgvs_predicted_protein_consequence.lrg_slr }}</a></td>
                                  <td><a href="{{ each.reference_sequence_records.lrg }}" target="_blank">{{ each.lrg_ac }}</a></td>
                              </tr>
                             {% endif %}
                          </table>
                          <h5>Genomic Variants</h5>
                          <table class="table table-bordered">
                              <tr>
                                  <th>Variant Description</th>
                                  <th>VCF Description</th>
                                  <th>Link to GenBank</th>
                              </tr>
                              {% if each.primary_assembly_loci.grch38.hgvs_genomic_description %}
                               <tr>
                                  <td><a class="variant-example" data-genome="GRCh38" href="#">{{ each.primary_assembly_loci.grch38.hgvs_genomic_description }}</a></td>
                                  <td>{{ each.primary_assembly_loci.grch38.vcfstr }}</td>
                                  <td><a href="https://www.ncbi.nlm.nih.gov/nuccore/{{ each.primary_assembly_loci.grch38.ac }}" target="_blank">{{ each.primary_assembly_loci.grch38.ac }}</a></td>
                              </tr>
                              {% endif %}
                              {% if each.primary_assembly_loci.grch37.hgvs_genomic_description %}
                              <tr>
                                  <td><a class="variant-example" data-genome="GRCh37" href="#">{{ each.primary_assembly_loci.grch37.hgvs_genomic_description }}</a></td>
                                  <td>{{ each.primary_assembly_loci.grch37.vcfstr }}</td>
                                  <td><a href="https://www.ncbi.nlm.nih.gov/nuccore/{{ each.primary_assembly_loci.grch37.ac }}" target="_blank">{{ each.primary_assembly_loci.grch37.ac }}</a></td>
                              </tr>
                              {% endif %}
                              {%  for alt in each.alt_genomic_loci %}
                                {% if alt.grch38 %}
                                  <tr>
                                      <td><a class="variant-example" data-genome={{ alt.grch38.genome }} href="#">{{ alt.grch38.hgvs_genomic_description }}</a></td>
                                      <td>{{ alt.grch38.vcfstr }}</td>
                                      <td><a href="https://www.ncbi.nlm.nih.gov/nuccore/{{ alt.grch38.ac }}" target="_blank">{{ alt.grch38.ac }}</a></td>
                                  </tr>
                                 {% endif %}
                                {% if alt.grch37 %}
                                  <tr>
                                      <td><a class="variant-example" data-genome={{ alt.grch37.genome }} href="#">{{ alt.grch37.hgvs_genomic_description }}</a></td>
                                      <td>{{ alt.grch37.vcfstr }}</td>
                                      <td><a href="https://www.ncbi.nlm.nih.gov/nuccore/{{ alt.grch37.ac }}" target="_blank">{{ alt.grch37.ac }}</a></td>
                                  </tr>
                                 {% endif %}
                              {% endfor %}
                          </table>
                          {% if ucsc %}
                              <h5>View Variants</h5>
                              <h6>View in <a href="{{ ucsc }}|{{ each.hgvs_transcript_variant }}" target="_blank">UCSC Genome Browser</a></h6>
                          {% endif %}
                          <br>
                          {% if each.variant_exonic_positions is not None %}
                              <h5>Variant Exon/Intron positions</h5>
                                  <table class="table table-bordered">
                                      <tr>
                                      <th>Genomic/Gene accession</th>
                                      <th>Exon/Intron at variant start</th>
                                      <th>Exon/Intron at variant end</th>
                                      </tr>
                                      {%  for genomic_ref, positions in  each.variant_exonic_positions.items %}
                                          <tr>
                                              <td>{{ genomic_ref }} </td>
                                              {% for key, val in positions.items %}
                                                  <td>{{ val }}</td>
                                              {%  endfor %}
                                          </tr>
                                      {% endfor %}
                                  </table>
                          {%  endif %}
                          <br>
                          <h4>Recommended Variant Descriptions</h4>
                          <ol>
                              <li>HGVS guidelines recommend using genomic and transcript descriptions in all publications</li>
                              <li>Use of the three- or one-letter amino acid alphabet is optional, but three-letter is recommended</li>
                          </ol>

                          <h5>Genomic descriptions</h5>
                          <table class="table table-bordered">
                              <tr>
                                  <th>Reference Sequence Type</th>
                                  <th>Variant Description</th>
                              </tr>
                              {% if each.hgvs_refseqgene_variant %}
                              <tr>
                                  <td>Gene</td>
                                  <td>{{ each.hgvs_refseqgene_variant }}</td>
                              </tr>
                              {%  endif %}
                              {% if each.primary_assembly_loci.grch37 %}
                              <tr>
                                  <td>Chromosomal GRCh37</td>
                                  <td>{{ each.primary_assembly_loci.grch37.hgvs_genomic_description }}</td>
                              </tr>
                              {%  endif %}
                              {% if each.primary_assembly_loci.grch38 %}
                              <tr>
                                  <td>Chromosomal GRCh38</td>
                                  <td>{{ each.primary_assembly_loci.grch38.hgvs_genomic_description }}</td>
                              </tr>
                              {%  endif %}
                          </table>
                          {% if not "Intergenic" in each.safe_hgvs_trans %}
                          {% if not each.rna_variant_descriptions.rna_variant %}
                          <h5>Transcript and protein descriptions</h5>
                          <table class="table table-bordered">
                              <tr>
                                  <th>Reference Sequence Type</th>
                                  <th>Variant Description</th>
                              </tr>
                              {% if each.genome_context_intronic_sequence %}
                              <tr>
                                  <td>Transcript</td>
                                  <td>{{ each.genome_context_intronic_sequence }}</td>
                              </tr>
                              {% endif %}
                              {% if each.refseqgene_context_intronic_sequence %}
                              <tr>
                                  <td>Transcript</td>
                                  <td>{{ each.refseqgene_context_intronic_sequence }}</td>
                              </tr>
                              {% endif %}
                              {% if not each.genome_context_intronic_sequence %}
                              <tr>
                                  <td>Transcript</td>
                                  <td>{{ each.hgvs_transcript_variant }}</td>
                              </tr>
                              {% endif %}
                              {% if each.hgvs_predicted_protein_consequence.slr  %}
                              <tr>
                                  <td>Protein single letter code</td>
                                  <td>{{ each.hgvs_predicted_protein_consequence.slr }}</td>
                              </tr>
                              {% endif %}
                              {% if each.hgvs_predicted_protein_consequence.tlr  %}
                                  <tr>
                                  <td>Protein three letter code</td>
                                  <td>{{ each.hgvs_predicted_protein_consequence.tlr }}</td>
                              </tr>
                              {% endif %}
                          </table>
                          {% endif %}

                          {% if each.rna_variant_descriptions.rna_variant %}
                          <h5>Transcript and protein descriptions</h5>
                          <table class="table table-bordered">
                              <tr>
                                  <th>Reference Sequence Type</th>
                                  <th>Variant Description</th>
                              </tr>
                              {% if each.rna_variant_descriptions.rna_variant %}
                              <tr>
                                  <td>Transcript</td>
                                  <td>{{ each.rna_variant_descriptions.rna_variant }}</td>
                              </tr>
                              {% endif %}
                              {% if each.rna_variant_descriptions.translation  %}
                              <tr>
                                  <td>Protein single letter code</td>
                                  <td>{{ each.rna_variant_descriptions.translation_slr }}</td>
                              </tr>
                              {% endif %}
                              {% if each.rna_variant_descriptions.translation %}
                                  <tr>
                                  <td>Protein three letter code</td>
                                  <td>{{ each.rna_variant_descriptions.translation }}</td>
                              </tr>
                              {% endif %}
                          </table>
                          {% endif %}

                          <br>
                          <h4>Gene Information</h4>
                          <table class="table table-bordered">
                              <tr>
                                  <th>Attribute</th>
                                  <th>Identifier</th>
                                  <th>Source</th>
                              </tr>
                              {% if each.gene_symbol %}
                              <tr>
                                  <td>Symbol</td>
                                  <td>{{ each.gene_symbol }}</td>
                                  <td><a href="https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/{{ each.gene_ids.hgnc_id }}" target="_blank">HGNC</a></td>
                              </tr>
                              {% endif %}
                              {% if each.annotations.note %}
                              <tr>
                                  <td>Name</td>
                                  <td>{{ each.annotations.note }}</td>
                                  <td><a href="https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/{{ each.gene_ids.hgnc_id }}" target="_blank">HGNC</a></td>
                              </tr>
                              {% endif %}
                              {% if 'HGNC:' in each.gene_ids.hgnc_id %}
                              <tr>
                                  <td>HGNC ID</td>
                                  <td>{{ each.gene_ids.hgnc_id }}</td>
                                  <td><a href="https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/{{ each.gene_ids.hgnc_id }}" target="_blank">HGNC</a></td>
                              </tr>
                              {% endif %}

                              {% if each.gene_ids.entrez_gene_id %}
                              <tr>
                                  <td>NCBI Gene ID</td>
                                  <td>{{ each.gene_ids.entrez_gene_id }}</td>
                                  <td><a href="https://www.ncbi.nlm.nih.gov/gene/{{ each.gene_ids.entrez_gene_id }}" target="_blank">NCBI</a></td>
                              </tr>
                              {%  endif %}

                              {% if each.gene_ids.ensembl_gene_id%}
                              <tr>
                                  <td>Ensembl Gene ID</td>
                                  <td>{{ each.gene_ids.ensembl_gene_id }}</td>
                                  <td><a href="https://www.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g={{ each.gene_ids.ensembl_gene_id }}" target="_blank">Ensembl</a></td>
                              </tr>
                              {%  endif %}
                                {% for omim in each.gene_ids.omim_id %}
                                  <tr>
                                    <td>OMIM ID</td>
                                    <td>{{ omim }}</td>
                                    <td><a href="https://www.omim.org/entry/{{ omim }}" target="_blank">OMIM</a></td>
                                  </tr>
                                {% endfor %}
                                {% for ccds in each.gene_ids.ccds_ids %}
                                  <tr>
                                    <td>CCDS ID</td>
                                    <td>{{ ccds }}</td>
                                    <td><a href="https://www.ncbi.nlm.nih.gov/projects/CCDS/CcdsBrowse.cgi?REQUEST=ALLFIELDS&DATA={{ ccds }}&ORGANISM=0&BUILDS=CURRENTBUILDS" target="_blank">CCDS</a></td>
                                  </tr>
                                {% endfor %}
                          </table>
                          {%  endif %}
                          {% if each.hgvs_transcript_variant %}
                              <form method="post" id="pdf-validate-form" name="pdf-validate-form">
                                  {% csrf_token %}
                                  <div class="form-group">
                                     {% if not each.rna_variant_descriptions.rna_variant %}
                                      <input type="hidden" id="variant_id_pdf" name="variant" type="text" class="form-control" required value="{{ each.hgvs_transcript_variant }}">
                                     {% endif %}
                                      {% if each.rna_variant_descriptions.rna_variant %}
                                      <input type="hidden" id="variant_id_pdf" name="variant" type="text" class="form-control" required value="{{ each.rna_variant_descriptions.rna_variant }}">
                                     {% endif %}
                                  </div>
                                  <div class="form-group" id="pdf-genomeselect" name="pdf-validate-form">
                                      <div class="custom-radio custom-control custom-control-inline">
                                          <input class="custom-control-input" type="radio" name="genomebuild" value={{ output.genome }}  checked>
                                      </div>
                                  </div>
                                  <div class="form-group" id="pdf-refsource" name="pdf-validate-form">
                                      <div class="custom-radio custom-control custom-control-inline">
                                          <input class="custom-control-input" type="radio" name="refsource" value={{ output.source }}  checked>
                                      </div>
                                  </div>
                                  <button id="pdf-validate-btn" type="submit" class="btn btn-primary" {% if locked %} disabled {% endif %}>
                                    Print pdf
                                  </button>
                              </form>
                          {% else %}
                              <form method="post" id="pdf-validate-form" name="pdf-validate-form">
                                  {% csrf_token %}
                                  <div class="form-group">
                                      {% if 'GRCh37' in output.genome%}
                                        <input type="hidden" id="variant_id_pdf" name="variant" type="text" class="form-control" required value="{{ each.primary_assembly_loci.grch37.hgvs_genomic_description }}">
                                      {% else %}
                                        <input type="hidden" id="variant_id_pdf" name="variant" type="text" class="form-control" required value="{{ each.primary_assembly_loci.grch38.hgvs_genomic_description }}">
                                      {% endif %}
                                  </div>
                                  <div class="form-group" id="pdf-genomeselect" name="pdf-validate-form">
                                      <div class="custom-radio custom-control custom-control-inline">
                                          <input class="custom-control-input" type="radio" name="genomebuild" value={{ output.genome }} checked>
                                      </div>
                                  </div>
                                  <div class="form-group" id="pdf-refsource" name="pdf-validate-form">
                                      <div class="custom-radio custom-control custom-control-inline">
                                          <input class="custom-control-input" type="radio" name="refsource" value={{ output.source }}  checked>
                                      </div>
                                  </div>
                                  <button id="pdf-validate-btn" type="submit" class="btn btn-primary" {% if locked %} disabled {% endif %}>
                                    Print pdf
                                  </button>
                              </form>
                          {% endif %}
                            <br>
                            <br>
                            <div class="row">
                              <div class="col-md-12 mb-5">
                                  <h4>Project variant onto other transcripts via a genomic position</h4>
                                      {% if each.primary_assembly_loci.grch38.hgvs_genomic_description %}
                                          <a class="variant-example" data-genome="GRCh38" href="#">{{ each.primary_assembly_loci.grch38.hgvs_genomic_description }}</a><p1> (GRCh38)</p1><br>
                                      {% endif %}
                                      {% if each.primary_assembly_loci.grch37.hgvs_genomic_description %}
                                          <a class="variant-example" data-genome="GRCh37" href="#">{{ each.primary_assembly_loci.grch37.hgvs_genomic_description }}</a><p1> (GRCh37)</p1><br>
                                      {% endif %}
                                      <p1>Alternatively, select a variant from the Genomic Variants table (above)</p1>
                               </div>
                            </div>
                      </div>
                  {% endfor %}
              </div>
          </div>
      </div>
      <div class="row">
          <div class="col-md-12 mb-5">

              <h4>View variant in external resources</h4>
               {%  if ucsc or varsome %}
               <ul>
                {% if varsome %}
                    <li><a href="{{ varsome }}" target="_blank">VarSome</a></li>
                {%  endif %}
                {% if gnomad %}
                    <li><a href="{{ gnomad }}" target="_blank">gnomAD_{{ output.genome }}</a></li>
                {% endif %}
              </ul>
              {%  endif %}
              {%  if not ucsc and not varsome and not gnomad %}
                  <li class="text-danger">No links to external resources available, possibly because the variant cannot be mapped to the selected genome build</li>
              {%  endif %}
           </div>
      </div>
  </div>
{% endif %}
<hr>
